### PR TITLE
Setup wizard can deploy Forseti to a shared VPC

### PR DESCRIPTION
**Pull request by bgleonard**
This PR will allow users who wish to run the Forseti Compute VM in a shared VPC to do so by passing arguments to the setup wizard. The user must specify the host project id, shared vpc name, and subnetwork name as arguments.

The generated compute instance template section networkInterfaces.network will now contain a host-project and vpc-name. The host-project can either be the host project id supplied by the user or the current project id as discovered by the script. The vpc-name can either be the vpc name supplied by the user or the name string 'default'.

The generated compute instance template section networkInterfaces.subnetwork is new. It will contain a host-project defined the same as above, region defined as it was previously, and subnetwork supplied by the user or the name string 'default'.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/forseti-security/841)

### CODE REVIEW

File: scripts/gcp_setup/environment/gcloud_env.py [hunk 1]
```diff
@@ -146,6 +146,9 @@ def __init__(self, **kwargs):
         self.notification_recipient_email = (
             kwargs.get('notification_recipient_email'))
         self.gsuite_superadmin_email = kwargs.get('gsuite_superadmin_email')
+        self.host_project_id = kwargs.get('host_project', self.project_id)
+        self.vpc_name = kwargs.get('vpc') or 'default'
```
Comment (blueandgold) [inline scripts/gcp_setup/environment/gcloud_env.py]: The get() would support a default option:
```
self.vpc_name = kwargs.get('vpc', 'default')
```

File: scripts/gcp_setup/environment/gcloud_env.py [hunk 2]
```diff
@@ -146,6 +146,9 @@ def __init__(self, **kwargs):
...
+        self.subnetwork = kwargs.get('subnet') or 'default'
```
Comment (blueandgold) [inline scripts/gcp_setup/environment/gcloud_env.py]: ditto as above:
```
self.subnetwork_name = kwargs.get('subnet_name', 'default')
```

File: scripts/gcp_setup/environment/gcloud_env.py [hunk 3]
```diff
@@ -159,6 +162,7 @@ def run_setup(self):
         self.get_organization()
         self.check_billing_enabled()
         self.has_permissions()
+        self.get_host_project()
```
Comment (blueandgold) [inline scripts/gcp_setup/environment/gcloud_env.py]: This method name might be better as ```check_network_host_project_id()```.  I know that other methods here also uses ```get``` naming, but I opened issue #842 to address

File: deployment-templates/compute-engine/forseti-instance.py.schema [hunk 4]
```diff
@@ -80,3 +80,15 @@ properties:
...
+    type: string
+    description: the id of the shared network host project
+
+  vpc-name:
+    type: string
+    description: the name of the shared network
+
+  subnetwork:
+    type: string
+    description: the name of the subnet
```
Comment (blueandgold) [inline deployment-templates/compute-engine/forseti-instance.py.schema]: add a new line

File: deployment-templates/compute-engine/forseti-instance.py.schema [hunk 5]
```diff
@@ -80,3 +80,15 @@ properties:
     type: integer
     default: 3306
     description: The port for the database.
+
+  network-host-project:
+    type: string
+    description: the id of the shared network host project
```
Comment (blueandgold) [inline deployment-templates/compute-engine/forseti-instance.py.schema]: For clarity, can you please rephrase this description:
```The id of the project that is hosting the shared network resources.```

File: deployment-templates/compute-engine/forseti-instance.py.schema [hunk 6]
```diff
@@ -80,3 +80,15 @@ properties:
...
+
+  vpc-name:
+    type: string
+    description: the name of the shared network
```
Comment (blueandgold) [inline deployment-templates/compute-engine/forseti-instance.py.schema]: Please capitalize ```The``` at the start of the description, and add a ```.``` at the end of the sentence.

File: deployment-templates/compute-engine/forseti-instance.py.schema (hunk 4 above)
Comment (blueandgold) [inline deployment-templates/compute-engine/forseti-instance.py.schema]: ditto: Please capitalize ```The``` at the start of the description, and add a ```.``` at the end of the sentence.

**Comment by blueandgold** [comment]
I also would like to see a unit test around whether the VPC or non-VPC deployment is handled correctly, but the best way to do that might involve testing the generated deployment template file, which is not in place and beyond the scope of this PR.  So, I have opened #843 to address separtely.

File: scripts/gcp_setup/environment/gcloud_env.py (hunk 1 above)
Comment (bgleonard) [reply scripts/gcp_setup/environment/gcloud_env.py]: I completely agree but this follows convention with the rest of this file (see line 135 for example). Happy to change it still.

File: deployment-templates/compute-engine/forseti-instance.py.schema (hunk 5 above)
Comment (bgleonard) [reply deployment-templates/compute-engine/forseti-instance.py.schema]: Complete

File: deployment-templates/compute-engine/forseti-instance.py.schema (hunk 6 above)
Comment (bgleonard) [reply deployment-templates/compute-engine/forseti-instance.py.schema]: Complete

File: deployment-templates/compute-engine/forseti-instance.py.schema (hunk 4 above)
Comment (bgleonard) [reply deployment-templates/compute-engine/forseti-instance.py.schema]: Complete

Comment (bgleonard) [reply deployment-templates/compute-engine/forseti-instance.py.schema]: Complete

File: deployment-templates/deploy-forseti.yaml.in [hunk 7]
```diff
@@ -53,13 +53,17 @@ resources:
...
+    region: $(ref.cloudsql-instance.region)
     zone: $(ref.cloudsql-instance.region)-c
 
     service-account: $(ref.forseti-gcp-reader.email)
     service-account-scopes:
       - https://www.googleapis.com/auth/cloud-platform
     scanner-bucket: {SCANNER_BUCKET}
     database-name: forseti_security
+    network-host-project: {HOST_PROJECT}
+    vpc-name: {XPN_NAME}
```
[the remark this answers is not in the document]
Comment (bgleonard) [reply deployment-templates/deploy-forseti.yaml.in]: Complete

File: scripts/gcp_setup/environment/gcloud_env.py (hunk 2 above)
Comment (bgleonard) [reply scripts/gcp_setup/environment/gcloud_env.py]: Same as above. I'll make the change if that's the direction the rest of the file is supposed to go.

File: scripts/gcp_setup/environment/gcloud_env.py (hunk 3 above)
Comment (bgleonard) [reply scripts/gcp_setup/environment/gcloud_env.py]: Complete

File: scripts/gcp_setup/environment/gcloud_env.py [hunk 8]
```diff
@@ -664,6 +675,9 @@ def generate_deployment_templates(self):
             'SERVICE_ACCT_GCP_READER': self.gcp_service_account,
             'SERVICE_ACCT_GSUITE_READER': self.gsuite_service_account,
             'BRANCH_OR_RELEASE': 'branch-name: "{}"'.format(self.branch),
+            'HOST_PROJECT': self.host_project_id,
+            'XPN_NAME': self.vpc_name,
```
[the remark this answers is not in the document]
Comment (bgleonard) [reply scripts/gcp_setup/environment/gcloud_env.py]: Complete

File: scripts/gcp_setup/setup_forseti.py [hunk 9]
```diff
@@ -40,6 +40,14 @@ def run():
     group.add_argument('--cloudsql-region',
                        help='The Cloud SQL region')
 
+    network = parser.add_argument_group(title='network')
+    network.add_argument('--host-project',
+                         help='The host project id')
```
[the remark this answers is not in the document]
Comment (bgleonard) [reply scripts/gcp_setup/setup_forseti.py]: Complete

File: scripts/gcp_setup/setup_forseti.py [hunk 10]
```diff
@@ -40,6 +40,14 @@ def run():
...
+    network.add_argument('--vpc',
```
[the remark this answers is not in the document]
Comment (bgleonard) [reply scripts/gcp_setup/setup_forseti.py]: Complete

File: scripts/gcp_setup/setup_forseti.py [hunk 11]
```diff
@@ -40,6 +40,14 @@ def run():
...
+                         help='The VPC name where Forseti VM will run')
+    network.add_argument('--subnet',
```
[the remark this answers is not in the document]
Comment (bgleonard) [reply scripts/gcp_setup/setup_forseti.py]: Complete

**Comment by blueandgold** [comment]
Thanks for making the changes!  Really awesome now!!  :lgtm_strong: 

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[scripts/gcp_setup/environment/gcloud_env.py, line 150 at r1](https://reviewable.io:443/reviews/googlecloudplatform/forseti-security/841#-Kzzg1oPA5f0uDPLGVi7-r1-150:-L-4FNV7iQPWHMUY0jvV:b-rxr3wy) ([raw file](https://github.com/googlecloudplatform/forseti-security/blob/e1a9fe92003d6af651a9da1ce3d14566e3d90288/scripts/gcp_setup/environment/gcloud_env.py#L150)):*
<details><summary><i>Previously, bgleonard wrote…</i></summary><blockquote>

I completely agree but this follows convention with the rest of this file (see line 135 for example). Happy to change it still.
</blockquote></details>

Good point.  I can make a sweep later, and clean all this up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/forseti-security/841#-:-L-4G2LuFJjEiYPl5SWy:b-95x7rp)*

File: scripts/gcp_setup/setup_forseti.py [hunk 12]
```diff
@@ -40,6 +40,15 @@ def run():
     group.add_argument('--cloudsql-region',
                        help='The Cloud SQL region')
 
+    network = parser.add_argument_group(title='network')
+    network.add_argument('--network-host-project-id',
```
Comment (blueandgold) [inline scripts/gcp_setup/setup_forseti.py]: I know this is a port from the previous PR, but I am wondering if we can take the chance to improve the naming?
`--vpc-host-project-id` ?

File: scripts/gcp_setup/setup_forseti.py [hunk 13]
```diff
@@ -40,6 +40,15 @@ def run():
...
+                         help='The project id that is hosting the network '
+                              'resources.')
+    network.add_argument('--vpc-name',
```
Comment (blueandgold) [inline scripts/gcp_setup/setup_forseti.py]: ``--vpc-host-network` ?

File: scripts/gcp_setup/setup_forseti.py [hunk 14]
```diff
@@ -40,6 +40,15 @@ def run():
...
+                         help='The VPC name where Forseti VM will run.')
+    network.add_argument('--subnet-name',
```
Comment (blueandgold) [inline scripts/gcp_setup/setup_forseti.py]: `--vpc-host-subnetwork` ?

**Pull request merged**